File: src/org/jgll/grammar/BodyGrammarSlot.java
package org.jgll.grammar;

import java.io.IOException;
import java.io.Serializable;
import java.io.Writer;

/**
 * 
 * @author Ali Afroozeh
 *
 */
public abstract class BodyGrammarSlot extends GrammarSlot implements Serializable {
	
	private static final long serialVersionUID = 1L;
	
	protected final Rule rule;
	
	protected final BodyGrammarSlot previous;
	
	protected BodyGrammarSlot next;
	
	private final String label;
	
	/**
	 * The position from the beginning of the alternate.
	 * Positions start from zero.
	 */
	protected final int position;
		
	public BodyGrammarSlot(Rule rule, int id, int position, BodyGrammarSlot previous) {
		super(id);
		this.position = position;
		this.rule = rule;
		if(previous != null) {
			previous.next = this;
		}
		this.previous = previous;
		
		
		// TODO: use getName() to generate the full slot presentation.
		String tmp = rule.getHead() + " ::= ";
		int i = 0;
		for(Symbol s : rule.getBody()) {
			if(i++ == position) {
				tmp += ". ";
			}
			tmp += s + " ";
		}
		if(position == rule.getBody().size()) {
			tmp += ".";
		}
		label = tmp;
	}
	
	/**
	 * Checks whether the provide input belogs to the first set, and follow set
	 * in case the first set contains epsilon.  
	 */
	public abstract boolean checkAgainstTestSet(int i);
	
	public abstract void codeIfTestSetCheck(Writer writer) throws IOException;
	
	public void codeElseTestSetCheck(Writer writer) throws IOException {
		writer.append("} else { newParseError(grammar.getGrammarSlot(" + this.id +  "), ci); label = L0; return; } \n");
	}
	
	public BodyGrammarSlot next() {
		return next;
	}
	
	public BodyGrammarSlot previous() {
		return previous;
	}
	
	/**
	 * Head ::= alpha . beta
	 * TODO: change this!
	 * 
	 * @return
	 */
	public Nonterminal getHead() {
		return rule.getHead();
	}
	
	public int getPosition() {
		return position;
	}
	
	@Override
	public String toString() {
		return label;
	}
	
	public abstract Iterable<Terminal> getTestSet();
}

File: src-scala/org/jgll_staged/grammar/slot/BodyGrammarSlot.scala
package org.jgll_staged.grammar.slot

import java.io.IOException
import java.io.Serializable
import java.io.Writer
import java.util.ArrayList
import java.util.List
import org.jgll_staged.grammar.Symbol
import org.jgll_staged.grammar.slotaction.SlotAction
import org.jgll_staged.parser.GLLParserInternals
import org.jgll_staged.util.Input
//remove if not needed
import scala.collection.JavaConversions._

@SerialVersionUID(1L)
abstract class BodyGrammarSlot(protected val position: Int, protected var _previous: BodyGrammarSlot, protected var head: HeadGrammarSlot)
    extends GrammarSlot with Serializable {

  protected var _next: BodyGrammarSlot = _

  protected var preConditions: List[SlotAction[Boolean]] = new ArrayList()

  var popActions: List[SlotAction[Boolean]] = new ArrayList()

  private var label: String = _

  if (position < 0) {
    throw new IllegalArgumentException("Position cannot be negative.")
  }

  if (_previous != null) {
    _previous._next = this
  }

  def addPopAction(popAction: SlotAction[Boolean]) {
    popActions.add(popAction)
  }

  def getPopActions(): java.lang.Iterable[SlotAction[Boolean]] = popActions

  def addPreCondition(preCondition: SlotAction[Boolean]) {
    preConditions.add(preCondition)
  }

  def getPreConditions(): List[SlotAction[Boolean]] = preConditions

  protected def executePreConditions(parser: GLLParserInternals, input: Input): Boolean = {
    preConditions.find(_.execute(parser, input)).isDefined
  }

  def testFirstSet(index: Int, input: Input): Boolean

  def testFollowSet(index: Int, input: Input): Boolean

  def codeIfTestSetCheck(writer: Writer): Unit

  def isFirst(): Boolean = _previous == null

  def codeElseTestSetCheck(writer: Writer) {
    writer.append("} else { newParseError(grammar.getGrammarSlot(" + this.id + 
      "), ci); label = L0; return; } \n")
  }

  def next(): BodyGrammarSlot = _next

  def previous(): BodyGrammarSlot = _previous

  def setPrevious(previous: BodyGrammarSlot) {
    this._previous = previous
  }

  def setNext(next: BodyGrammarSlot) {
    this._next = next
  }

  def getPosition(): Int = position

  def getHead(): HeadGrammarSlot = head

  def getSymbol(): Symbol

  def isNullable(): Boolean

  def isNameEqual(slot: BodyGrammarSlot): Boolean

  def setLabel(label: String) {
    this.label = label
  }

  override def toString(): String = label
}
